### PR TITLE
feat: add sign permission confirmation for erc20 revocation permission

### PR DESCRIPTION
**Pull request by jeffsmale90**
## **Description**

Adds confirmation content for the `erc20-token-revocation` permission type.

Because this is the first permission type that is not a "token spend" type permission, it requires a different title, so this PR also adds this differentiation. Presently this is a simple switch statement. If the collection of titles grows to an unwieldy number, we can decompose it into a helper function, but I've left in place for simplicity.

This PR does _not_ introduce the new permission type. That will be done in a future change that introduces updated permission snaps. (hence the no-changelog labe).

Also updates controllers released in https://github.com/MetaMask/core/pull/7330

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/38495?quickstart=1)

## **Changelog**

CHANGELOG entry: null

## **Manual testing steps**

Test in conjunction with https://github.com/MetaMask/snap-7715-permissions/pull/234

## **Screenshots/Recordings**

image

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

> [!NOTE]
> Adds UI/i18n and handling for `erc20-token-revocation` confirmations, updates selectors/tests, adjusts LavaMoat policies, and bumps related controller dependencies.
> 
> - **Confirmations UI**:
>   - Add `erc20-token-revocation` support to typed-sign flow with `Erc20TokenRevocationDetails` and integration in `typed-sign-permission.tsx`.
>   - Introduce `getPermissionDescription` to show a distinct title/description for revocation vs. spend permissions.
>   - Minor util refactor to accept an i18n function type.
> - **i18n**:
>   - New message key `confirmTitleDescERC20Revocation` in `app/_locales/en*/messages.json`.
> - **Gator Permissions**:
>   - Include `erc20-token-revocation` in hooks/selectors test fixtures and edge cases.
> - **Policies**:
>   - Update LavaMoat policy to reference `@metamask/signature-controller` directly and remove nested alias entries across builds.
> - **Dependencies**:
>   - Bump `@metamask/shield-controller` to `^4.0.0`, `@metamask/signature-controller` to `^38.0.0`, `@metamask/gator-permissions-controller` to `^0.8.0` (and transitive updates like `transaction-controller` and `logging-controller`).
> - **Tests**:
>   - Add tests for ERC-20 revocation details; update existing tests to include the new permission type.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 037f40cf7ee3d81dc020b153464983fdd2d7aafa. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: ui/pages/confirmations/components/confirm/info/typed-sign/typed-sign-permission/erc20-token-revocation-details.tsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * @returns JSX element containing the ERC20 token revocation permission details
+ */
+export const Erc20TokenRevocationDetails: React.FC<{
+  expiry: number | null;
+}> = ({ expiry }) => {
+  const t = useI18nContext();
+
+  return (
+    <ConfirmInfoSection data-testid="erc20-token-revocation-details-section">
+      {expiry && (
```
Comment (OGPoyraz) [inline ui/pages/confirmations/components/confirm/info/typed-sign/typed-sign-permission/erc20-token-revocation-details.tsx]: Can we early `return null` instead of checking `expiry` inline? 

Because if it doesn't exists then it will put an empty div (caused by `ConfirmInfoSection`) with some classes which can cause layout to be broken.

Comment (jeffsmale90) [reply ui/pages/confirmations/components/confirm/info/typed-sign/typed-sign-permission/erc20-token-revocation-details.tsx]: 👍 

Good call - I've made this change as you suggest.

**Review by OGPoyraz — DISMISSED** [review]

**Comment by jeffsmale90** [comment]
Sorry for the force push here - I ended up going around in circles trying to get the lavamoat policies updated. Made for a much cleaner PR if I collated the changes afresh.

File: ui/pages/confirmations/components/confirm/title/title.tsx [hunk 2]
```diff
@@ -203,7 +204,13 @@ const getDescription = (
         return t('confirmTitleDescPermitSignature');
       }
       if ((confirmation as SignatureRequestType).decodedPermission) {
-        return t('confirmTitleDescPermission');
+        const permissionType = (confirmation as SignatureRequestType)
+          .decodedPermission?.permission?.type;
+
+        return getPermissionDescription(
+          t as ReturnType<typeof useI18nContext>,
```
Comment (mj-kiwi) [inline ui/pages/confirmations/components/confirm/title/title.tsx]: `t` is already typed as `IntlFunction`. Should we redefine `IntlFunction` as an alias of `ReturnType<typeof useI18nContext>` to keep things consistent?

**Review by mj-kiwi — COMMENTED** [review]
LGTM. I just have one question: the number of tests for erc20-token-revocation is much lower than the other permission types. Should we address this in a future PR, or handle it in this one? 
I think we need to update the `CHANGELOG entry: ` as well

**Review by mj-kiwi — DISMISSED** [review]
LGTM

**Review by OGPoyraz — DISMISSED** [review]

**Review by OGPoyraz — APPROVED** [review]

**Review by mj-kiwi — APPROVED** [review]

**Review by Gudahtt — APPROVED** [review]
Policy changes look good 👍

File: lavamoat/browserify/beta/policy.json [hunk 3]
```diff
@@ -1993,7 +1993,7 @@
       "packages": {
         "@metamask/base-controller": true,
         "@metamask/controller-utils": true,
-        "@metamask/shield-controller>@metamask/signature-controller": true,
+        "@metamask/signature-controller": true,
```
Comment (Prithpal-Sooriya) [inline lavamoat/browserify/beta/policy.json:1996]: Pathing change, acceptable.

File: lavamoat/browserify/beta/policy.json [hunk 4]
```diff
@@ -2018,24 +2018,6 @@
         "uuid": true
       }
     },
-    "@metamask/shield-controller>@metamask/signature-controller": {
```
Comment (Prithpal-Sooriya) [inline lavamoat/browserify/beta/policy.json:2021]: Peer deps have been moved to direct dependencies from controller bump ([changelog](https://github.com/MetaMask/core/blob/main/packages/shield-controller/CHANGELOG.md#310)) - looks good.

**Review by Prithpal-Sooriya — APPROVED** [review]
